presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/MapSqlFunctions.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.scalar.sql;
+
+import com.facebook.presto.spi.function.Description;
+import com.facebook.presto.spi.function.SqlInvokedScalarFunction;
+import com.facebook.presto.spi.function.SqlParameter;
+import com.facebook.presto.spi.function.SqlParameters;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.function.TypeParameter;
+
+public class MapSqlFunctions
+{
+    private MapSqlFunctions() {}
+
+    @SqlInvokedScalarFunction(value = "map_top_n_keys", deterministic = true, calledOnNullInput = false)
+    @Description("Returns the top N keys of the given map in descending order according to the natural ordering of its values.")
+    @TypeParameter("K")
+    @TypeParameter("V")
+    @SqlParameters({@SqlParameter(name = "input", type = "map(K, V)"), @SqlParameter(name = "n", type = "bigint")})
+    @SqlType("array<K>")
+    public static String mapTopNKeys()
+    {
+        return "RETURN IF(n >= 0, slice(reverse(array_sort(map_keys(input))), 1, n), slice(array_sort(map_keys(input)), 1, ABS(n)))";
+    }
+
+    @SqlInvokedScalarFunction(value = "map_top_n_keys", deterministic = true, calledOnNullInput = true)
+    @Description("Returns the top N keys of the given map in sorted using the provided lambda comparator.")
+    @TypeParameter("K")
+    @TypeParameter("V")
+    @SqlParameters({@SqlParameter(name = "input", type = "map(K, V)"), @SqlParameter(name = "n", type = "bigint"), @SqlParameter(name = "f", type = "function(K, K, int)")})
+    @SqlType("array<K>")
+    public static String mapTopNKeysComparator()
+    {
+        return "RETURN IF(n >= 0, slice(reverse(array_sort(map_keys(input), f)), 1, n), slice(array_sort(map_keys(input), f), 1, ABS(n)))";
+    }
```

Comment:
Also you don't need the ABS etc., slice works both ways - if you give negative index it goes reverse. so just slice(.. ,1, n) should do it.

Reply:
I think in slice negative index still just set the pointer from reverse direction but the slice still gives you n elements starting from that element

Reply:
> I think in slice negative index still just set the pointer from reverse direction but the slice still gives you n elements starting from that element

yeah - in that case start with -N?

Reply:
So maybe just do:

SLICE(..., LEAST(1, -N), N)

Reply:
Hmm...I can do this but not sure if it will have any advantage over what I'm already doing. So using the above method for N > 0, I still need to reverse the array. For example

SELECT SLICE(ARRAY[1, 2, 3, 4], LEAST(1, -3), 3) -> output: [2, 3, 4 ], expected: [4, 3, 2]

Reply:
> Hmm...I can do this but not sure if it will have any advantage over what I'm already doing. So using the above method for N > 0, I still need to reverse the array. For example
> 
> SELECT SLICE(ARRAY[1, 2, 3, 4], LEAST(1, -3), 3) -> output: [2, 3, 4 ], expected: [4, 3, 2]

Is that the expected? I'm not sure. In fact, I'm beginning to think we disallow N < 0. Let them define a comparator that reverses the direction. That saves a lot of confusion. So I would do:

IF(N < 0, fail('N must be non-negative'), ...)

Reply:
Also let's update the documentation to say if they want bottom N, they need a comparator that reverses the direction

Reply:
We can disallow negative N but then every time if need bottom N keys even for simple key data types, people do need to pass the lambda. But yeah I am happy to throw error in case N <= 0.  Let me know

Reply:
That's fine. We need clean semantics in the language and sometimes that makes things hard for some use cases. We can always add MAP_BOTTOM_N_KEYS as a separate udf if needed - especially since we now have SQL UDFs supported better.

Reply:
Cool!